### PR TITLE
level-1/task 2

### PR DESCRIPTION
**Pull request by PolinaXox**
Task 2. First try.
There is no nice "professional" comments. There are some "my_own_for_studing_only_comments".

### CODE REVIEW

**Comment by PolinaXox** [comment]
Task 2. 2nd try. 
Code's cleaned and coments've added by rules.

File: 1. Basics/task_2/hw2.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * @return string the input request example as a string
+ */
+function readHttpLikeInput() : string {
+    $f = fopen( 'php://stdin', 'r' );
+    $store = "";
+    $toread = 0;
+    while( $line = fgets( $f ) ) {
```
Comment (anton-moroz) [inline 1. Basics/task_2/hw2.php:9]: empty line before

File: 1. Basics/task_2/hw2.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * @return string the input request example as a string
```
Comment (anton-moroz) [inline 1. Basics/task_2/hw2.php:3]: put description above return

File: 1. Basics/task_2/hw2.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    $strAsArr = explode("\n", $string);
+    $methodAndURI = explode(" ", array_shift($strAsArr));
+    $headers = [];
+    $body = NULL;
+
+    // if the request has headers store each header data as
+    // $headers[headerName] = headerValue
+    if($strAsArr) {
+        $i = 0;
+        while($strAsArr[$i]) {
```
Comment (anton-moroz) [inline 1. Basics/task_2/hw2.php:47]: empty line

File: 1. Basics/task_2/hw2.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            $ind = strpos($strAsArr[$i], ":");
```
Comment (anton-moroz) [inline 1. Basics/task_2/hw2.php:48]: try to simplify this logic
there is a simpler way to do it.

File: 1. Basics/task_2/hw2.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            $headers[trim(substr($strAsArr[$i], 0, $ind))] = trim(substr($strAsArr[$i], $ind+1));
```
Comment (anton-moroz) [inline 1. Basics/task_2/hw2.php:49]: if there is no `:` in the line (in body for example) - you will have incorrect behavior and errors.
if `strpos` doesn't find `:` - it returns `false`. The key will be empty string, but the value will be not "string", but "tring", as `false + 1 = 1`

File: 1. Basics/task_2/hw2.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            $i++;
+        }
+
+        // empyty line between header(s) and body, if it is
+        $i++;
+        
+        // if the request has a body
```
Comment (anton-moroz) [inline 1. Basics/task_2/hw2.php:56]: you will not get here and body will be always empty

**Review by anton-moroz — CHANGES_REQUESTED** [review]

File: 1. Basics/task_2/hw2.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    if ($toread > 0) 
+        $store .= fread($f, $toread);
+    return $store;
+}
+
+$contents = readHttpLikeInput();
+
+// parsing 
+function parseTcpStringAsHttpRequest($string) {
+    $strAsArr = explode("\n", $string);
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php:38]: There was a problem: my win PHP_EOL equals "\r\n" and param of the function contains (if contains) only one "\r\n" between the headers and the body.
But I won it with super fuction
``` php
function getStrToParse($string) : string {
    $str = preg_replace("/\r/", "", $string);
    $str = preg_replace("/\n/", PHP_EOL, $str);
    return $str;
}

```

File: 1. Basics/task_2/hw2.php [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    $methodAndURI = explode(" ", array_shift($strAsArr));
+    
+    $headers = array();
```
[the remark this answers is not in the document]
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php]: Done.

File: 1. Basics/task_2/hw2.php (hunk 2 above)
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php:3]: Done

File: 1. Basics/task_2/hw2.php (hunk 3 above)
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php:47]: Done.

File: 1. Basics/task_2/hw2.php (hunk 6 above)
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php:56]: All sourse I saw tells that there have to be an empty line between the header and the body if there is a body.
But I've found different solution. But it seems like something not about http requests and responses necessary structure. 
But it was hard, and funny, and took a loooot of time.

File: 1. Basics/task_2/hw2.php (hunk 4 above)
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php]: Done.

File: 1. Basics/task_2/hw2.php [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+ * 
+ * @param $string
+ * 
+ * @return string
+ */
+
+function getStrToParse($string) : string {
+    $str = preg_replace("/\r/", "", $string);
+    $str = preg_replace("/\n/", PHP_EOL, $str);
+    return $str;
```
Comment (anton-moroz) [inline 1. Basics/task_2/hw2.php:95]: code styles

File: 1. Basics/task_2/hw2.php (hunk 5 above)
Comment (anton-moroz) [reply 1. Basics/task_2/hw2.php]: you use strpos too much. in this case there is an easier way to achieve what you want

File: 1. Basics/task_2/hw2.php (hunk 6 above)
Comment (anton-moroz) [reply 1. Basics/task_2/hw2.php]: but you managed to do it. Nice work!

**Review by anton-moroz — CHANGES_REQUESTED** [review]

File: 1. Basics/task_2/hw2.php (hunk 5 above)
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php]: Do you mean split by ":" with ```explode()```? Or using regex` function?
All options are good for nice headers.
For bad header: "HelloIAmHeader: IHaveSomeMore_:_thanYouExpect ":
* strpos() will show junk,
* regex like ```'/^(.*):(.*)$/'``` may show junk,
*  expload() without some other checking will not show junk but smth like good one if 
``` php
$arr = explode(":", $badHeader);
$headers += [$arr[0] => $arr[1]];
```
It's better to see the junk after bad requests processig.

I guessed which option you mean?))

File: 1. Basics/task_2/hw2.php (hunk 9 above)
Comment (PolinaXox) [reply 1. Basics/task_2/hw2.php:95]: Done.
...if you mean that ```return``` is a block...

**Review by anton-moroz — APPROVED** [review]

**Pull request merged**